sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,684 @@
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.
+Interfaces defined in this specification may not be implemented yet or may be
+in a preliminary state.
+The specification itself may also change in incompatible ways before it is
+finalized.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+This extension introduces a new way to define a kernel as a simple C++
+function, where the kernel arguments are parameters to the function.
+This is different from standard SYCL kernels, where the kernel arguments are
+either captures of a lambda expression or member variables of a callable
+object.
+
+The primary motivation for this extension is the
+link:../proposed/sycl_ext_oneapi_kernel_compiler.asciidoc[
+sycl_ext_oneapi_kernel_compiler], which allows online compilation of a kernel
+from source code.
+These kernels must have a clean separation between host and device code, and
+they must have a defined order to their arguments because the application sets
+the argument values by their "index" via `handler::set_arg(index, value)`.
+Because variables captured by lambda expressions have no defined order, we need
+some other way to define kernel arguments for use with the
+sycl_ext_oneapi_kernel_compiler extension, and the
+sycl_ext_oneapi_free_function_kernels extension provides that mechanism.
+The "free function kernel" feature is a separate extension, though, because it
+can also be used independently from sycl_ext_oneapi_kernel_compiler.
+For example, some users may find it more familiar to define kernels as plain
+functions or they might have other reasons to prefer a clean separation between
+host and device code.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_FREE_FUNCTION_KERNELS`
+to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Defining a free function kernel
+
+A free function kernel is a normal C++ function definition, where the function
+declaration is decorated with one of the following macros:
+`SYCL_EXT_ONEAPI_RANGE_KERNEL`, `SYCL_EXT_ONEAPI_ND_RANGE_KERNEL`, or
+`SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL`.
+
+When a function declaration is decorated with one of these macros, the
+following rules must be observed:
+
+* The function must be declared at either namespace scope are in class scope as
+  a static member function.
```

Comment:
```suggestion
* The function must be declared at either namespace scope or at class scope as
  a static member function.
```
In order to support integration headers, I think the function must also be restricted to having a return type and parameter types that are forward declarable at namespace scope (e.g., they can't be member class types or enumeration types without an explicit underlying type).

Reply:
Thanks for catching the typo.

Yes, I agree that this isn't sufficient to support integration headers.  I'm working on a separate list of restrictions for that.  My thought is to write the main part of the specification assuming an implementation that does not rely on integration headers.  (E.g. it might be a single-pass implementation or an implementation that does integration at IR level.)  Then, I will have a separate section that lists additional limitations for implementation that use integration headers.

The restriction I added here (7d9b6f60bd1cf707ace7b08510a3c90cd74e6166) is a general requirement that applies to all implementations.